### ISSUE TITLE
Log messages split between files on log rotation

### ISSUE THREAD

**Issue by argerus** [rsyslog/rsyslog#409]
When upgrading from 5.8.11 -> 8.10.0 we've noticed that log messages started to get split up between files during log rotation.

```
# tail -n3 /var/log/info.log.2
2015-06-23T06:01:07.337+00:00 q1614 [ INFO    ] stclient[698]: [avhs-dbus.c:99] avhs service has name 'com.axis.AVHS'
2015-06-23T06:01:07.339+00:00 q1614 [ INFO    ] [avhs-dbus.c:111] object path '/com/axis/AVHS' already registered
2015-06-23T06:01:07.340+00:00 q1614 [ INFO    ] stclient[698]: [av
```
```
# head -n3 /var/log/info.log.1
hs-dbus.c:111] object path '/com/axis/AVHS' already registered
2015-06-23T06:01:07.377+00:00 q1614 [ INFO    ] Started Remote Service Client Daemon.
2015-06-23T06:01:07.574+00:00 q1614 [ INFO    ] UPnP Initialized(192.168.0.198:49152)
```

Log rotation is handled by out channels in our rsyslog config, and since this is an embedded application we are quite size restricted:

```
$outchannel INFO,/var/log/info.log,9000,/usr/share/rsyslog/log_rotateinfo.log.sh
$outchannel WARNING,/var/log/warning.log,5333,/usr/share/rsyslog/log_rotatewarning.log.sh
$outchannel ERROR,/var/log/error.log,4000,/usr/share/rsyslog/log_rotateerror.log.sh
$outchannel CRITICAL,/var/log/critical.log,3000,/usr/share/rsyslog/log_rotatecritical.log.sh

*.info;*.!warning;kern.none;authpriv.none;auth.none       :omfile:$INFO
*.=warning;kern.none;authpriv.none;auth.none              :omfile:$WARNING
*.=err;kern.none;authpriv.none;auth.none                  :omfile:$ERROR
*.crit;kern.none;authpriv.none;auth.none                  :omfile:$CRITICAL
```

My testing tells me the bug was introduced with the 7.5.4 release.

Looking at the diffs from 7.5.3 I noticed that a couple of default values where changed, which might be what caused this bug to surface. If I understand the source code correctly, using `$ResetConfigVariables` resets variables to their "legacy values" as opposed to their default values (is that on purpose or a bug?), and when I put that at the top of our rsyslog.conf the issue disappears. But I guess the underlying bug remains.

Further testing suggests (re-)setting only `$MainMsgQueueDequeueBatchSize 32` (which was one of the defaults that was changed in 7.5.4) makes the issue disappear as well (without using `$ResetConfigVariables`).

**Comment by argerus** [comment]
Even further testing suggests this issue was present in 7.5.3 as well, but I only notice it after manually setting `$MainMsgQueueDequeueBatchSize 256` in the config.

**Comment by alorbach** [comment]
Have you tried Build 8.11.0? 
I tried to reproduce this but without success. 

Please post a full working configuration example, this would make it much easier to reproduce the problem under the right circumstances.

**Comment by argerus** [comment]
Hi, back from vacation now.

>Have you tried Build 8.11.0?

Yes, I've tried 8.11.0 and master as well. The bug seems to only occur during boot on our embedded system, which is a time when both CPU and IO experiences some heavy load. I think I have been able to reproduce it more consequently after boot as well by adding a sleep to our log rotation scripts. This reproduced to the issue not only on our embedded system but on my developer machine aswell.

[This gist](https://gist.github.com/argerus/0aa0277888eabf28719b) contains the log rotation script and test script as well as the changes I made to the stock Debian rsyslog.conf (Wheezy, with rsyslog 5.8.11), to reproduce it.

Yeah and about that. I noticed that the issue is present even before 7.3.2 if `$RepeatedMsgReduction` is set to `off` (which is the default). In our embedded config this was set to `on` which masked the issue until 7.3.2 when changes where made to the main loop regarding flushing messages and `$RepeatedMsgReduction`.

**Comment by argerus** [comment]
Have you been able to reproduce the issue with the additional files given in my gist above?

**Comment by jer-gentoo** [comment]
Yes, it seems like right after starting, rsyslogd does not respond to HUP properly (yet) in that it doesn't close+reopen omfile file handles at that time.

**Comment by rgerhards** [comment]
This could be related to https://github.com/rsyslog/rsyslog/pull/554 . Maybe this patch fixes that situation here, too.

**Comment by rgerhards** [comment]
this should be fixed in 8.16.0. if not, please re-open.
